### ISSUE TITLE
get permissions for the app

### ISSUE THREAD

**Issue by hongkongkiwi** [facundoolano/google-play-scraper#20]
I would love to get permissions from the app, it's possible to get by using the web interface.

**Comment by ankit-m** [comment]
Same here. It would be great if permissions could be added.

**Comment by facundoolano** [comment]
The permissions are not loaded on the first request to the app page but on a separate API call, which I couldn't make work outside google play after a couple of quick tests. 

The endpoint is `https://play.google.com/store/xhr/getdoc`. If anyone manages to get it working, let me know and I'll add a permissions method to the module.

**Comment by ankit-m** [comment]
I could make it work, but I cannot filter out the permissions from the response. It is a simple post request (with session in python) to `https://play.google.com/store/xhr/getdoc?authuser=0`. Add the 'ids' as the name of the app and 'xhr' to 1. 

The response doesn't look like JSON (there a lot of lists nested inside lists). But the permissions are in the last list. If you could help with the filtering.

**Comment by facundoolano** [comment]
Ok, I've managed to get a response with this request:

```
curl --data "ids=com.dxco.pandavszombies&xhr=1" https://play.google.com/store/xhr/getdoc?authuser=0
```

And yes, the response is a mess. I'll play a bit with it when I have some time and see if I can parse the permissions.

**Comment by facundoolano** [comment]
I didn't find a straight forward way to parse the response above. Closing this for now.

**Comment by MarcelloLins** [comment]
@facundoolano  Have you managed to find a way to parse the permissions properly ?

**Comment by facundoolano** [comment]
Haven't tried again. Any input is welcome.

**Comment by MarcelloLins** [comment]
@facundoolano  I have been bashing my head against it for the past 2 hours already. I'ts hard to find a pattern, but it gotta have one, otherwise not even Google would be able to parse it back to it's rendereable form.

I also couldn't find the script responsible for rendering the permissions. If I could find it, to reverse it would't be that hard.

**Comment by MarcelloLins** [comment]
@facundoolano I managed to put something that is really close to 100% working right here : [PermissionsParserFile](https://github.com/MarcelloLins/GooglePlayAppsCrawler/blob/master/SharedLibrary/PlayStoreParser.cs).

It uses the http response that comes from the HTTP Post request for the "Permissions Popup Box" and extracts the permissions out of it.

**Comment by facundoolano** [comment]
Interesting, I'll see if I can port it to this project. Thanks!

**Comment by id0Sch** [comment]
hi, i've been following this thread for a while since i'm in need for permissions as well.
the we do it is to actually download the apk and parse the manifest - that way we get the permissions consts used by android.

@MarcelloLins's request also returns banners - @facundoolano, do you think you will be able to incorporate banners to this lib as well?

thanks, and great job guys!

**Comment by MarcelloLins** [comment]
Hello @id0Sch . Do you mind sharing how do you manage to download the APKs in an automated way, in scale ? 

I would love to understand how exactly that works behind the scenes

**Comment by id0Sch** [comment]
https://github.com/dweinstein/node-google-play

**Comment by MarcelloLins** [comment]
@id0Sch I don't have the guts to debug Node.JS/ JS code. I really don't. I could have been ANY other language and I would deep dive in it, but not NODE hahahaha

I will pass on that, thanks though!

**Comment by facundoolano** [comment]
@id0Sch what do you mean by banners? can you show me an example in google play?

**Comment by id0Sch** [comment]
@MarcelloLins , what it does is to login to googlePlay with the given android device and credentials and issue the request to a) get app data b) download apk

@facundoolano, when you use 
```
curl --data "ids=com.dxco.pandavszombies&xhr=1" https://play.google.com/store/xhr/getdoc?authuser=0
```
look for that line (should be directly under the description)
```
,[2,,,,"https://lh4.ggpht.com/kKfRICvVTCikV4MLqsP0kWEth2F-I1Qt4jxMdklOdE2r8AmtrE-Umn6_WH_cGExXnjk-",,"",,1,,,,,,,,,,,,,0]
```
it means creative type 2, which is banner

**Comment by facundoolano** [comment]
Thanks for the pointers everyone. I've added the [permissions method](https://github.com/facundoolano/google-play-scraper#permissions) in the latest version.

**Comment by stanislasdrg** [comment]
Asking for an app permissions now returns an error 
`TypeError: Cannot read property '42656262' of null` The error points to permissions.js, line 49 :
`return res[0][2][0][60]['42656262'][1] || [];`

**Comment by facundoolano** [comment]
I imagined this day would come :P 

Will need to review that funky response and see how it changed

**Comment by stanislasdrg** [comment]
Had a look at the code and response, you just need to change to 65 instead of 60.
That is : `return res[0][2][0][65]['42656262'][1] || [];`

It's indeed the funkiest response I've ever seen. Any ideas on why they would randomly change the response ?

**Comment by facundoolano** [comment]
I don't know that whole payload is a mystery, no idea where those numbers come from.

Have you tested that fix? Can you send a PR?

**Comment by stanislasdrg** [comment]
Yep tested, I get the proper response. Will send a PR.

**Comment by facundoolano** [comment]
awesome, thanks!

**Comment by facundoolano** [comment]
fixed by #126
